core/node/vm_runner/src/impls/bwip.rs

```diff
@@ -0,0 +1,380 @@
+            Self { vm_runner },
+            BasicWitnessInputProducerTasks {
+                loader_task,
+                output_handler_factory_task,
+            },
+        ))
+    }
+
+    /// Continuously loads new available batches and writes the corresponding data
+    /// produced by that batch.
+    ///
+    /// # Errors
+    ///
+    /// Propagates RocksDB and Postgres errors.
+    pub async fn run(self, stop_receiver: &watch::Receiver<bool>) -> anyhow::Result<()> {
+        self.vm_runner.run(stop_receiver).await
+    }
+}
+
+/// A collections of tasks that need to be run in order for BWIP to work as
+/// intended.
+#[derive(Debug)]
+pub struct BasicWitnessInputProducerTasks {
+    /// Task that synchronizes storage with new available batches.
+    pub loader_task: StorageSyncTask<BasicWitnessInputProducerIo>,
+    /// Task that handles output from processed batches.
+    pub output_handler_factory_task:
+        ConcurrentOutputHandlerFactoryTask<BasicWitnessInputProducerIo>,
+}
+
+#[derive(Debug, Clone)]
+pub struct BasicWitnessInputProducerIo {
+    first_processed_batch: L1BatchNumber,
+    window_size: u32,
+}
+
+#[async_trait]
+impl VmRunnerIo for BasicWitnessInputProducerIo {
+    fn name(&self) -> &'static str {
+        "basic_witness_input_producer"
+    }
+
+    async fn latest_processed_batch(
+        &self,
+        conn: &mut Connection<'_, Core>,
+    ) -> anyhow::Result<L1BatchNumber> {
+        Ok(conn
+            .vm_runner_dal()
+            .get_bwip_latest_processed_batch()
+            .await?
+            .unwrap_or(self.first_processed_batch))
+    }
+
+    async fn last_ready_to_be_loaded_batch(
+        &self,
+        conn: &mut Connection<'_, Core>,
+    ) -> anyhow::Result<L1BatchNumber> {
+        Ok(conn
+            .vm_runner_dal()
+            .get_bwip_last_ready_batch(self.first_processed_batch, self.window_size)
+            .await?)
+    }
+
+    async fn mark_l1_batch_as_completed(
+        &self,
+        conn: &mut Connection<'_, Core>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<()> {
+        Ok(conn
+            .vm_runner_dal()
+            .mark_bwip_batch_as_completed(l1_batch_number)
+            .await?)
+    }
+}
+
+#[derive(Debug)]
+struct BasicWitnessInputProducerOutputHandler {
+    pool: ConnectionPool<Core>,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+#[async_trait]
+impl StateKeeperOutputHandler for BasicWitnessInputProducerOutputHandler {
+    async fn handle_l2_block(&mut self, _updates_manager: &UpdatesManager) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    async fn handle_l1_batch(
+        &mut self,
+        updates_manager: Arc<UpdatesManager>,
+    ) -> anyhow::Result<()> {
+        let l1_batch_number = updates_manager.l1_batch.number;
+        let mut connection = self.pool.connection().await?;
+
+        tracing::info!(
+            "Started saving VM run data for L1 batch {:?}",
+            l1_batch_number
+        );
```

Comment:
I don't see it changed, leaving suggestion
```suggestion
        tracing::info!(%l1_batch_number, "Started saving VM run data");
```

Reply:
Ah, I didn't understand you correctly, I see now, thanks